lib/ui/painting/gradient.cc

```diff
@@ -46,28 +47,32 @@ void CanvasGradient::initLinear(const tonic::Float32List& end_points,
   SkPoint p0 = SkPoint::Make(end_points[0], end_points[1]);
   SkPoint p1 = SkPoint::Make(end_points[2], end_points[3]);
   std::vector<DlColor> dl_colors;
-  dl_colors.reserve(colors.num_elements());
-  for (int i = 0; i < colors.num_elements(); ++i) {
-    /// TODO(gaaclarke): Make this preserve wide gamut colors.
-    dl_colors.emplace_back(DlColor(colors[i]));
+  dl_colors.reserve(num_colors);
```

Comment:
Can we have a helper method, or would that involve unnecessary copying? I thought the compiler optimized the case of returning a vector...??

Another way is to create the vector on the local stack and then hand it into the converter method by reference.

Reply:
> Can we have a helper method, or would that involve unnecessary copying? I thought the compiler optimized the case of returning a vector...??
> 
> Another way is to create the vector on the local stack and then hand it into the converter method by reference.

You can't have a vector on a stack.  It will always allocate memory on the heap.  The size is needed to be known at compilation time in order to put something on the stack (without trickery).  I have a proposition in an above thread for using an `std::array` on the stack.

I think avoiding the copying will require DlColorSource some mechanism to read from the tonic collection directly.

Reply:
The data has to be heap allocated because it is growable, that will happen whether a helper function is used or not. I'm referring to the object itself, but I suppose that the object isn't very big whether it is stack allocated or not. It's a vtable + data pointer + allocated size + used size.

Reply:
Oh, I see, sorry I thought this was still related to the heap discussion.  Yes we could have a helper function and yes C++ would perform the return value optimization.